src/main/scala/workflow/graph/DanglingNodeRemovalRule.scala

```diff
@@ -0,0 +1,23 @@
+package workflow.graph
+
+/**
+ * A rule to remove all nodes & sources in a graph that don't lead to any sink.
+ */
+object DanglingNodeRemovalRule extends Rule {
+  override def apply(plan: Graph, prefixes: Map[NodeId, Prefix]): (Graph, Map[NodeId, Prefix]) = {
+    val ancestorsOfSinks = plan.sinks.foldLeft(Set[GraphId]()) {
+      case (ancestors, sink) => ancestors ++ AnalysisUtils.getAncestors(plan, sink)
+    }
+
+    val nodesToRemove = plan.nodes -- ancestorsOfSinks.collect { case node: NodeId => node }
+    val sourcesToRemove = plan.sources -- ancestorsOfSinks.collect { case source: SourceId => source }
+
+    val afterSourceRemoval = sourcesToRemove.foldLeft(plan) {
+      case (curPlan, sourceToRemove) => curPlan.removeSource(sourceToRemove)
+    }
+
+    nodesToRemove.foldLeft((afterSourceRemoval, prefixes)) {
+      case ((curPlan, curPrefixes), nodeToRemove) => (curPlan.removeNode(nodeToRemove), curPrefixes - nodeToRemove)
```

Comment:
I'm a little worried that maintaining this prefix state is going to come back to bit us, but I'm willing to live with it.